src/store/api.ts

```diff
@@ -300,3 +310,71 @@ export const apiMiddleware: Middleware = ({ getState }) => (next: Dispatch<AnyAc
 
   return next(action)
 }
+
+// utility to auto refresh token
+// TODO(chrishwong): if there are multiple simultaneous API calls, there will be
+// multiple calls to refresh the token. We should add some handling to pause
+// subsequent requests while the refresh is happening
+export const maybeRefreshToken = async (token: string, refreshToken: string, dispatch): Promise<any> => {
+  try {
+    if (token) {
+      var decoded: TokenClaims = jwtDecode(token)
+      var exp = decoded.exp
+      var now = Math.floor(Date.now() / 1000) // in seconds
+      if ((exp - now) < 300) { // we preemptively refresh up to 5 min early for smoother operation
+        let res = await refreshSession(token, refreshToken)
+
+        // fire an action to update the access token in state
+        dispatch({
+          type: 'SET_TOKEN',
+          token: res.data.access_token
+        })
+
+        return Promise.resolve(res.data.access_token)
+      }
+    }
+  } catch (error) {
+    console.log(`error refreshing token - ${error.message}`)
+    if (token) {
+      return Promise.resolve(token)
+    }
+    // maybe invalidate session?
```

Comment:
Do we have logic for re-routing to the login page if the user is on or tries to navigate to a private route without having a valid token? If so, this is where we should reroute them, or at least, catch this error higher up in the chain & send them to the login page.

Reply:
Nice catch, I will work this in